src/main/java/frc/robot/scoreStates/ShootState.java

```diff
@@ -0,0 +1,63 @@
+package frc.robot.scoreStates;
+
+import com.pathplanner.lib.auto.AutoBuilder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Commands;
+import frc.robot.Constants;
+import frc.robot.swerve.SwerveDrive;
+import java.util.List;
+import java.util.Set;
+import lib.Utils;
+
+public class ShootState implements ScoreState {
+
+    @Override
+    public Command driveToClosestOptimalPoint() {
+        return Commands.defer(
+                () -> {
+                    Translation2d speakerPose;
+                    Pose2d botPose = SwerveDrive.getInstance().getBotPose();
+                    List<Translation2d> optimalPoints;
+                    if (isRed()) {
+                        speakerPose = ScoreStateConstants.SPEAKER_POSE_RED;
+                        optimalPoints = ScoreStateConstants.OPTIMAL_POINTS_SHOOT_RED;
+                    } else {
+                        speakerPose = ScoreStateConstants.SPEAKER_POSE_BLUE;
+                        optimalPoints = ScoreStateConstants.OPTIMAL_POINTS_SHOOT_BLUE;
+                    }
+                    Translation2d optimalTranslation;
+                    if (Utils.getDistanceFromPoint(speakerPose, botPose)
+                            < ScoreStateConstants.MAX_SHOOTING_DISTANCE.in(Units.Meters)) {
+                        optimalTranslation = botPose.getTranslation();
+                        return SwerveDrive.getInstance()
+                                .turnCommand(
+                                        Utils.calculateOptimalRotation(
+                                                        optimalTranslation, speakerPose)
+                                                .getRotations(),
+                                        ScoreStateConstants.TURN_TOLERANCE);
+                    }
+                    optimalTranslation = botPose.getTranslation().nearest(optimalPoints);
+
+                    return AutoBuilder.pathfindToPose(
+                            new Pose2d(
+                                    optimalTranslation,
+                                    Utils.calculateOptimalRotation(
+                                            optimalTranslation, speakerPose)),
+                            Constants.AUTO_CONSTRAINTS);
```

Comment:
This doesn't handle the rotation. You could use a `turnCommand` after pathfinding, but that wastes time. Better to use the "pathfind then follow path" feature.

Reply:
Pathfind then follow path will not do anything because you can not only rotate in a path created on the fly

Reply:
You have to rotate in place because you want to shoot from where you are

Reply:
I might be misexplaining something. I suggest something like this:
![image](https://github.com/Galaxia5987/robot-2024/assets/31829093/5de2ab7f-e893-493b-803b-a05351921dd2)
The red path is "pathfound", then the blue one is not. You rotate to the desired rotation while following the second path.

Something I realized that this condition might be too simple:
https://github.com/Galaxia5987/robot-2024/blob/61991aa618dd108de722763f52bbca38e6b38529/src/main/java/frc/robot/scoreStates/ShootState.java#L32-L33
For example, if we are under the Stage's Core, we might be within shooting distance, but we can't actually shoot from there.

Reply:
There is no actual difference in what you suggested from what currently happens, because if the robot is that far it will have time to rotate via pathfinding anyway.

Reply:
Nevermind. I remembered pathfinding couldn't control the heading, but confused "heading" with "rotation". Resolving.